File: sdks/python/insforge/http.py
"""
Core HTTP client for InsForge SDK.
Handles authentication, request building, and response parsing.
"""

from __future__ import annotations

import threading
from typing import Any
from urllib.parse import urljoin

import requests
from requests import Response, Session


class InsForgeError(Exception):
    """Raised when the InsForge API returns an error response."""

    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        error_code: str | None = None,
        next_actions: str | None = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.error_code = error_code
        self.next_actions = next_actions

    def __repr__(self) -> str:
        return (
            f"InsForgeError(message={self.message!r}, "
            f"status_code={self.status_code}, "
            f"error_code={self.error_code!r})"
        )


class _SessionState:
    """Thread-safe in-memory session storage."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._access_token: str | None = None
        self._refresh_token: str | None = None
        self._csrf_token: str | None = None
        self._user: dict[str, Any] | None = None

    def set(
        self,
        *,
        access_token: str | None = None,
        refresh_token: str | None = None,
        csrf_token: str | None = None,
        user: dict[str, Any] | None = None,
    ) -> None:
        with self._lock:
            if access_token is not None:
                self._access_token = access_token
            if refresh_token is not None:
                self._refresh_token = refresh_token
            if csrf_token is not None:
                self._csrf_token = csrf_token
            if user is not None:
                self._user = user

    def clear(self) -> None:
        with self._lock:
            self._access_token = None
            self._refresh_token = None
            self._csrf_token = None
            self._user = None

    @property
    def access_token(self) -> str | None:
        with self._lock:
            return self._access_token

    @property
    def refresh_token(self) -> str | None:
        with self._lock:
            return self._refresh_token

    @property
    def csrf_token(self) -> str | None:
        with self._lock:
            return self._csrf_token

    @property
    def user(self) -> dict[str, Any] | None:
        with self._lock:
            return self._user


class HttpClient:
    """Low-level HTTP client used by all SDK modules."""

    def __init__(
        self,
        base_url: str,
        anon_key: str,
        edge_function_token: str | None = None,
        timeout: int = 30,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.anon_key = anon_key
        self.edge_function_token = edge_function_token
        self.timeout = timeout
        self._session = Session()
        self._state = _SessionState()

        if edge_function_token:
            self._state.set(access_token=edge_function_token)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _get_auth_token(self) -> str:
        """Return the best available bearer token."""
        return self._state.access_token or self.anon_key

    def _build_headers(self, extra: dict[str, str] | None = None) -> dict[str, str]:
        headers: dict[str, str] = {
            "Authorization": f"Bearer {self._get_auth_token()}",
            "Content-Type": "application/json",
        }
        if extra:
            headers.update(extra)
        return headers

    @staticmethod
    def _raise_for_error(response: Response) -> None:
        if response.ok:
            return
        try:
            body = response.json()
        except Exception:
            body = {}
        raise InsForgeError(
            message=body.get("message", response.text or "Unknown error"),
            status_code=response.status_code,
            error_code=body.get("error"),
            next_actions=body.get("nextActions"),
        )

    def _url(self, path: str) -> str:
        return urljoin(self.base_url + "/", path.lstrip("/"))

    # ------------------------------------------------------------------
    # HTTP verbs
    # ------------------------------------------------------------------

    def get(
        self,
        path: str,
        params: dict[str, Any] | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> Any:
        r = self._session.get(
            self._url(path),
            params=params,
            headers=self._build_headers(extra_headers),
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        return r.json() if r.content else None

    def post(
        self,
        path: str,
        data: Any = None,
        params: dict[str, Any] | None = None,
        extra_headers: dict[str, str] | None = None,
        raw: bool = False,
    ) -> Any:
        headers = self._build_headers(extra_headers)
        r = self._session.post(
            self._url(path),
            json=data,
            params=params,
            headers=headers,
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        if raw:
            return r
        return r.json() if r.content else None

    def patch(
        self,
        path: str,
        data: Any = None,
        params: dict[str, Any] | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> Any:
        r = self._session.patch(
            self._url(path),
            json=data,
            params=params,
            headers=self._build_headers(extra_headers),
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        return r.json() if r.content else None

    def put(
        self,
        path: str,
        data: Any = None,
        extra_headers: dict[str, str] | None = None,
        files: dict[str, Any] | None = None,
    ) -> Any:
        headers = self._build_headers(extra_headers)
        if files:
            # multipart upload — drop Content-Type so requests sets boundary
            headers.pop("Content-Type", None)
            r = self._session.put(
                self._url(path),
                files=files,
                headers=headers,
                timeout=self.timeout,
            )
        else:
            r = self._session.put(
                self._url(path),
                json=data,
                headers=headers,
                timeout=self.timeout,
            )
        self._raise_for_error(r)
        return r.json() if r.content else None

    def delete(
        self,
        path: str,
        params: dict[str, Any] | None = None,
        data: Any = None,
        extra_headers: dict[str, str] | None = None,
    ) -> Any:
        r = self._session.delete(
            self._url(path),
            json=data,
            params=params,
            headers=self._build_headers(extra_headers),
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        return r.json() if r.content else None

    def post_multipart(
        self,
        path: str,
        files: dict[str, Any],
        data: dict[str, Any] | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> Any:
        headers = self._build_headers(extra_headers)
        headers.pop("Content-Type", None)  # let requests set multipart boundary
        r = self._session.post(
            self._url(path),
            files=files,
            data=data,
            headers=headers,
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        return r.json() if r.content else None

    def get_raw(
        self,
        path: str,
        extra_headers: dict[str, str] | None = None,
    ) -> bytes:
        r = self._session.get(
            self._url(path),
            headers=self._build_headers(extra_headers),
            timeout=self.timeout,
        )
        self._raise_for_error(r)
        return r.content

    def post_external(
        self,
        url: str,
        files: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> Response:
        """POST to an external URL (e.g., S3 presigned upload)."""
        r = self._session.post(url, files=files, data=data, headers=headers, timeout=self.timeout)
        r.raise_for_status()
        return r

    def put_external(
        self,
        url: str,
        content: bytes,
        content_type: str,
    ) -> None:
        """PUT raw bytes to an external URL."""
        r = self._session.put(
            url,
            data=content,
            headers={"Content-Type": content_type},
            timeout=self.timeout,
        )
        r.raise_for_status()

File: sdks/go/insforge/http.go
package insforge

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"path"
	"strings"
	"sync"
	"time"
)

// InsForgeError is returned when the InsForge API responds with an error.
type InsForgeError struct {
	Message    string `json:"message"`
	StatusCode int    `json:"statusCode"`
	ErrorCode  string `json:"error"`
	NextActions string `json:"nextActions"`
}

func (e *InsForgeError) Error() string {
	if e.ErrorCode != "" {
		return fmt.Sprintf("insforge: [%s] %s (HTTP %d)", e.ErrorCode, e.Message, e.StatusCode)
	}
	return fmt.Sprintf("insforge: %s (HTTP %d)", e.Message, e.StatusCode)
}

// sessionState holds the in-memory auth tokens. Thread-safe.
type sessionState struct {
	mu           sync.RWMutex
	accessToken  string
	refreshToken string
	csrfToken    string
	user         map[string]any
}

func (s *sessionState) setTokens(access, refresh, csrf string, user map[string]any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if access != "" {
		s.accessToken = access
	}
	if refresh != "" {
		s.refreshToken = refresh
	}
	if csrf != "" {
		s.csrfToken = csrf
	}
	if user != nil {
		s.user = user
	}
}

func (s *sessionState) clear() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.accessToken = ""
	s.refreshToken = ""
	s.csrfToken = ""
	s.user = nil
}

func (s *sessionState) getAccessToken() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.accessToken
}

func (s *sessionState) getRefreshToken() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.refreshToken
}

// httpClient is the low-level HTTP transport used by all SDK modules.
type httpClient struct {
	baseURL string
	anonKey string
	timeout time.Duration
	state   *sessionState
	client  *http.Client
}

func newHTTPClient(baseURL, anonKey, edgeFunctionToken string, timeoutSecs int) *httpClient {
	state := &sessionState{}
	if edgeFunctionToken != "" {
		state.accessToken = edgeFunctionToken
	}
	return &httpClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		anonKey: anonKey,
		timeout: time.Duration(timeoutSecs) * time.Second,
		state:   state,
		client:  &http.Client{Timeout: time.Duration(timeoutSecs) * time.Second},
	}
}

func (h *httpClient) bearerToken() string {
	if t := h.state.getAccessToken(); t != "" {
		return t
	}
	return h.anonKey
}

func (h *httpClient) buildURL(p string) string {
	base, _ := url.Parse(h.baseURL)
	base.Path = path.Join(base.Path, p)
	return base.String()
}

func (h *httpClient) newRequest(method, urlStr string, body io.Reader) (*http.Request, error) {
	req, err := http.NewRequest(method, urlStr, body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+h.bearerToken())
	req.Header.Set("Content-Type", "application/json")
	return req, nil
}

func (h *httpClient) parseError(resp *http.Response) error {
	var apiErr InsForgeError
	body, _ := io.ReadAll(resp.Body)
	if err := json.Unmarshal(body, &apiErr); err != nil {
		apiErr.Message = string(body)
	}
	apiErr.StatusCode = resp.StatusCode
	return &apiErr
}

// do executes a request, checks for errors, and decodes the response into out.
// Pass nil for out to discard the response body.
func (h *httpClient) do(method, p string, reqBody any, queryParams url.Values, out any, extraHeaders map[string]string) error {
	var bodyReader io.Reader
	if reqBody != nil {
		b, err := json.Marshal(reqBody)
		if err != nil {
			return fmt.Errorf("insforge: marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(b)
	}

	urlStr := h.buildURL(p)
	if len(queryParams) > 0 {
		urlStr += "?" + queryParams.Encode()
	}

	req, err := h.newRequest(method, urlStr, bodyReader)
	if err != nil {
		return err
	}
	for k, v := range extraHeaders {
		req.Header.Set(k, v)
	}

	resp, err := h.client.Do(req)
	if err != nil {
		return fmt.Errorf("insforge: http: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return h.parseError(resp)
	}

	if out == nil {
		return nil
	}
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("insforge: read response: %w", err)
	}
	if len(respBody) == 0 {
		return nil
	}
	return json.Unmarshal(respBody, out)
}

// doRaw executes a request and returns the raw response body bytes.
func (h *httpClient) doRaw(method, p string, reqBody any, queryParams url.Values, extraHeaders map[string]string) ([]byte, error) {
	var bodyReader io.Reader
	if reqBody != nil {
		b, err := json.Marshal(reqBody)
		if err != nil {
			return nil, err
		}
		bodyReader = bytes.NewReader(b)
	}

	urlStr := h.buildURL(p)
	if len(queryParams) > 0 {
		urlStr += "?" + queryParams.Encode()
	}

	req, err := h.newRequest(method, urlStr, bodyReader)
	if err != nil {
		return nil, err
	}
	for k, v := range extraHeaders {
		req.Header.Set(k, v)
	}

	resp, err := h.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return nil, h.parseError(resp)
	}
	return io.ReadAll(resp.Body)
}

// doStream executes a request and returns the response body as a streaming reader.
// The caller is responsible for closing the reader.
func (h *httpClient) doStream(method, p string, reqBody any) (io.ReadCloser, error) {
	var bodyReader io.Reader
	if reqBody != nil {
		b, err := json.Marshal(reqBody)
		if err != nil {
			return nil, err
		}
		bodyReader = bytes.NewReader(b)
	}

	req, err := h.newRequest(method, h.buildURL(p), bodyReader)
	if err != nil {
		return nil, err
	}

	// Use a client without timeout for streaming
	streamClient := &http.Client{}
	resp, err := streamClient.Do(req)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		defer resp.Body.Close()
		return nil, h.parseError(resp)
	}
	return resp.Body, nil
}

// doMultipart performs a multipart/form-data POST to an InsForge path.
func (h *httpClient) doMultipart(p string, fields map[string]string, fileField, filename string, fileContent []byte, contentType string, out any) error {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)

	for k, v := range fields {
		_ = w.WriteField(k, v)
	}

	if fileField != "" {
		fw, err := w.CreateFormFile(fileField, filename)
		if err != nil {
			return err
		}
		if _, err = fw.Write(fileContent); err != nil {
			return err
		}
	}
	w.Close()

	req, err := http.NewRequest(http.MethodPost, h.buildURL(p), &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+h.bearerToken())
	req.Header.Set("Content-Type", w.FormDataContentType())

	resp, err := h.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return h.parseError(resp)
	}
	if out == nil {
		return nil
	}
	respBody, _ := io.ReadAll(resp.Body)
	if len(respBody) == 0 {
		return nil
	}
	return json.Unmarshal(respBody, out)
}

// doPutExternal PUTs raw bytes to an external URL (e.g., presigned S3).
func (h *httpClient) doPutExternal(externalURL string, data []byte, contentType string) error {
	req, err := http.NewRequest(http.MethodPut, externalURL, bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", contentType)
	resp, err := h.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("insforge: external upload failed (HTTP %d): %s", resp.StatusCode, string(body))
	}
	return nil
}

// doPostExternal POSTs multipart form data to an external URL (presigned S3 POST).
func (h *httpClient) doPostExternal(externalURL string, fields map[string]string, fileKey, filename string, fileData []byte, contentType string) error {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)
	for k, v := range fields {
		_ = w.WriteField(k, v)
	}
	fw, err := w.CreateFormFile(fileKey, filename)
	if err != nil {
		return err
	}
	if _, err = fw.Write(fileData); err != nil {
		return err
	}
	w.Close()

	req, err := http.NewRequest(http.MethodPost, externalURL, &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", w.FormDataContentType())

	resp, err := h.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("insforge: external upload failed (HTTP %d): %s", resp.StatusCode, string(body))
	}
	return nil
}
